### ISSUE TITLE
Renaming is not atomic, crash can lead to data loss

### ISSUE THREAD

**Issue by pcworld** [NVSL/linux-nova#116]
Renaming on NOVA appears not to be atomic from a crash consistency perspective, and probably also not from a visibility perspective (from other threads).

For example, when files `a` and `b` exist and one renames `b` to `a`, two intermediate states can be observed before the final, correct state: 1) only `b` exists (`b` points to the old file and there is no file which points to the contents of `a`), 2) both `a` and `b` are missing.

This can be reproduced as follows (tested on commit 0de30c540bcc24a33c2e69735777c6a0aaddfbfb):

## Case 1 (only `b` exists)
Apply the following patch (returns after [the first `nova_remove_dentry`](https://github.com/NVSL/linux-nova/blob/0de30c540bcc24a33c2e69735777c6a0aaddfbfb/fs/nova/namei.c#L768)):
```diff
diff --git a/fs/nova/namei.c b/fs/nova/namei.c
index 4d3953091873..301a8002bd33 100644
--- a/fs/nova/namei.c
+++ b/fs/nova/namei.c
@@ -767,6 +767,7 @@ static int nova_rename(struct inode *old_dir,
                /* First remove the old entry in the new directory */
                err = nova_remove_dentry(new_dentry, 0, &update_dir_new,
                                        epoch_id);
+               goto out;
                if (err)
                        goto out;
        }
```

The following shell output demonstrates a crash at the early return we have patched in:
```
/ # set -e
/ # mount -tNOVA -oinit /dev/pmem0 /mnt
/ # echo -n tes1 > /mnt/myfile
/ # echo -n tes2 > /mnt/myfile2
/ # ls -lah /mnt
total 9
drwxr-xr-x    2 0        0           4.0K Oct 19 14:41 .
drwxr-xr-x   13 0        0              0 Oct 19 14:41 ..
-rw-r--r--    1 0        0              4 Oct 19 14:41 myfile
-rw-r--r--    1 0        0              4 Oct 19 14:41 myfile2
/ # sync
/ # mv /mnt/myfile2 /mnt/myfile
/ # ls -lah /mnt
total 5
drwxr-xr-x    2 0        0           4.0K Oct 19 14:41 .
drwxr-xr-x   13 0        0              0 Oct 19 14:41 ..
-rw-r--r--    1 0        0              4 Oct 19 14:41 myfile2
# Now simulate a crash
/ # cp /dev/pmem0 /bak
/ # umount /mnt
/ # cat /bak > /dev/pmem0
/ # mount -tNOVA /dev/pmem0 /mnt
/ # ls -lah /mnt 
total 5
drwxr-xr-x    2 0        0           4.0K Oct 19 14:41 .
drwxr-xr-x   13 0        0              0 Oct 19 14:44 ..
-rw-r--r--    1 0        0              4 Oct 19 14:41 myfile2
```
Note that even before simulating a crash, there is no file named `myfile`. It seems likely that this is also observable from other  threads, and thus not only a crash consistency but also a visibility problem, but I have not verified this.

## Case 2 (both `a` and `b` are missing)
Apply the following patch (returns after the [second `nova_remove_dentry`](https://github.com/NVSL/linux-nova/blob/0de30c540bcc24a33c2e69735777c6a0aaddfbfb/fs/nova/namei.c#L790)):
```diff
diff --git a/fs/nova/namei.c b/fs/nova/namei.c
index 4d3953091873..1794ce0f1c3b 100644
--- a/fs/nova/namei.c
+++ b/fs/nova/namei.c
@@ -789,6 +789,7 @@ static int nova_rename(struct inode *old_dir,
 
        err = nova_remove_dentry(old_dentry, dec_link, &update_dir_old,
                                        epoch_id);
+       goto out;
        if (err)
                goto out;
 
```
The following shell output demonstrates a crash at the early return we have patched in:
```
/ # set -e
/ # mount -tNOVA -oinit /dev/pmem0 /mnt
/ # echo -n tes1 > /mnt/myfile
/ # echo -n tes2 > /mnt/myfile2
/ # ls -lah /mnt
total 9
drwxr-xr-x    2 0        0           4.0K Oct 19 14:51 .
drwxr-xr-x   13 0        0              0 Oct 19 14:50 ..
-rw-r--r--    1 0        0              4 Oct 19 14:51 myfile
-rw-r--r--    1 0        0              4 Oct 19 14:51 myfile2
/ # sync
/ # mv /mnt/myfile2 /mnt/myfile
/ # ls -lah /mnt
total 1
drwxr-xr-x    2 0        0           4.0K Oct 19 14:51 .
drwxr-xr-x   13 0        0              0 Oct 19 14:50 ..
# Now simulate a crash
/ # cp /dev/pmem0 /bak
/ # umount /mnt
/ # cat /bak > /dev/pmem0
/ # mount -tNOVA /dev/pmem0 /mnt
/ # ls -lah /mnt
total 1
drwxr-xr-x    2 0        0           4.0K Oct 19 14:51 .
drwxr-xr-x   13 0        0              0 Oct 19 14:51 ..
```
As in case 1, the bug appears to manifest even without crashing.

## Other manifestations
- The same issue applies when renaming a file to a not-yet-existent file name (the file is lost in between).  
  (`mount -tNOVA -oinit /dev/pmem0 /mnt && echo -n tes2 > /mnt/myfile2 && sync && mv /mnt/myfile2 /mnt/myfile && ls -lah /mnt`)
- Even whole directory trees can be lost.  
  (`mount -tNOVA -oinit /dev/pmem0 /mnt && mkdir /mnt/newdir && echo -n test > /mnt/newdir/testfile && sync && mv /mnt/newdir /mnt/newdir2 && ls -lahR /mnt`)

## Violations
If it is true that the intermediate states are visible from other threads as I suspect, then this is at least a violation of [POSIX guarantees of `rename`](https://pubs.opengroup.org/onlinepubs/9699919799.2018edition/) (emphasis mine):
> If the old argument points to the pathname of a file that is not a directory, the new argument shall not point to the pathname of a directory. If the link named by the new argument exists, it shall be removed and old renamed to new. In this case, **a link named new shall remain visible to other threads throughout the renaming operation** and refer either to the file referred to by new or old before the operation began

While POSIX does not explicitly specify crash guarantees of the rename operation [0], this behavior appears unexpected as it can result in data loss.
Also, the NOVA paper explicitly writes about atomicity of rename operations (although a bit ambiguous whether meant only for visibility or also crash consistency), and the journaling mechanism used for that.
[At the end of `nova_rename`](https://github.com/NVSL/linux-nova/blob/0de30c540bcc24a33c2e69735777c6a0aaddfbfb/fs/nova/namei.c#L823-L852), there is some journaling going on. I'm not sure what exactly it's doing, so either it's for other special cases, or it does not work in the intended way (because as shown above, crashing directly after the `nova_remove_dentry` operations makes these removals directly visible).

[0]: [Specifying and Checking File System Crash-Consistency Models](https://unsat.cs.washington.edu/papers/bornholt-ferrite.pdf) by Bornholt et al.

**Comment by hayley-leblanc** [comment]
I have also encountered this issue. I've been investigating the root cause and I think I've found two issues in the code that contribute to it. Based on some preliminary testing on my end, fixing these issues fixes the rename atomicity issue in the default configuration of NOVA; I haven't tried it the checksum or parity modes, but I will be doing so soon. Both issues stem from modifications to the dentry being deleted not being properly journaled. The line numbers given here might be a bit off from the mainline repo; I think there might be some debugging print statements still sitting around in my local version. 

First, `nova_rename()` uses `nova_remove_dentry()` to delete the old dentry on line 790 of namei.c. By default, this is done in place. However, it shouldn't be done in place, because this results in the old version of the file being deleted too early. I initially tried to fix this by disabling in-place dentry deletion entirely, but this broke some other things. I believe the solution is to just ensure that inplace deletion is skipped for this specific call to `nova_remove_dentry()`; I made this change locally and tested it, and it resolved some of the issues.

Some crash states are still weird after this fix. Specifically, if we're renaming foo to bar, there are a couple crash states in which foo is still listed by `ls`, but `stat`-ing it fails with ENOENT. I believe this issue stems from the fact that foo's dentry's `reassigned` field is not set until after the rename transaction has committed; if we crash after the transaction commits but before the `reassigned` field is set, then `foo` will be incorrectly listed in its old directory. I modified `nova_create_rename_transaction()` to include the relevant dentry in the transaction and to ensure that the dentry's `reassigned` field is updated as part of the transaction, and this seems to fix the problem. 

I plan to do a bit more testing with this fix, but I can make a PR for review if these fixes seem reasonable.
